chartsync: support Git HTTPS credentials using secret

richardcase opened the pull request:
Started to add https basic authentication to the git chart source. the
username and password can be set as a static value or from a secret.
For example:

```yaml
spec:
  releaseName: my-app
  chart:
    git: https://github.com/someonesomerepo.git
    ref: master
    path: charts/my-app
    secretRef:
      name: git-credentials

```

Still a work in progress

Relates to: #94

richardcase committed:
feat: adding https basic auth to git source

Started to add https basic authentication to the git chart source. the
username and password can be set as a static value or from a secret.
For example:

```yaml
spec:
  releaseName: my-app
  chart:
    git: https://github.com/someonesomerepo.git
    ref: master
    path: charts/my-app
    auth:
      username:
        value: joebloggs
      password:
        valueFrom:
          name: git-pwd
          key: authpwd

```

Relates to: #94

richardcase committed:
chore: added error handling and comments

Added error handling/logging if there is an error reported when
adding the auth details. Also added comments to new types.

stefanprodan commented:
@richardcase I think this should be paused until we decide how to handle git/helm repos outside the HelmRelease spec. See https://github.com/fluxcd/helm-operator/issues/142

richardcase commented:
> @richardcase I think this should be paused until we decide how to handle git/helm repos outside the HelmRelease spec. See #142

@stefanprodan  - Sure. Although i really need this functionality for a specific scenario i'm working on. I'll work from a fork. If you need any help with #142 let me know

stefanprodan commented:
If you need this then please make it more compact like in my proposal:

```yaml
spec:
  releaseName: my-app
  chart:
    git: https://github.com/someonesomerepo.git
    ref: master
    path: charts/my-app
    secretRef: #optional
      name: git-basic-auth #required
```

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: git-basic-auth
type: Opaque
data:
  username: <BASE64> #required
  password: <BASE64> #required
```

richardcase commented:
> If you need this then please make it more compact like in my proposal

Sure i can do that. I will make that change now

hiddeco commented:
In addition to all of the above, instead of merging in `master` it is better for reviewers and Git history clutter to rebase on top of `master`.

richardcase commented:
> In addition to all of the above, instead of merging in `master` it is better for reviewers and Git history clutter to rebase on top of `master`.

Understood, that was a mess up my side....a case of PEBKAC

richardcase committed:
feat: https auth definition simpler

Based on review feedback the way to specify the http auth credentials
has been simplified so that you just provide a secret name. The
referenced secret is assumed to have 2 keys: username, password.

richardcase commented:
@stefanprodan @hiddeco  - i've made the change to simplify the specifiying of the creds as suggested. So as suggested you have:

```yaml
spec:
  releaseName: my-release
  chart:
    git: https://github.com/someorg/somerepo.git
    ref: master
    path: charts/my-chart
    secretRef:
      name: git-pwd
```
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: git-pwd
type: Opaque
data:
  password: supersecret
  username: joeblogs
```

hiddeco commented on cmd/helm-operator/main.go:
```diff
@@ -23,8 +23,8 @@ import (
-	"github.com/fluxcd/helm-operator/pkg/helm/v2"
-	"github.com/fluxcd/helm-operator/pkg/helm/v3"
+	v2 "github.com/fluxcd/helm-operator/pkg/helm/v2"
+	v3 "github.com/fluxcd/helm-operator/pkg/helm/v3"
```
The alias here is obsolete, and is probably caused by some enabled linter.

richardcase replied:
Changed this.

hiddeco commented on pkg/apis/helm.fluxcd.io/v1/types.go:
```diff
@@ -9,6 +9,7 @@ import (
 	"github.com/ghodss/yaml"
 
 	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
These are duplicates (`corev1` should be kept).

richardcase replied:
Changed

hiddeco commented on pkg/chartsync/git.go:
```diff
@@ -3,16 +3,22 @@ package chartsync
 	"k8s.io/client-go/tools/cache"
 
-	"github.com/fluxcd/helm-operator/pkg/apis/helm.fluxcd.io/v1"
+	v1 "github.com/fluxcd/helm-operator/pkg/apis/helm.fluxcd.io/v1"
```
Same as the above.

richardcase replied:
@hiddeco - every time i removed the v1 alias it adds it back so this hasn't been resolved i'm afraid

hiddeco replied:
I will remove it for you before the merge, no worries.

hiddeco commented on pkg/chartsync/git.go:
```diff
@@ -3,16 +3,22 @@ package chartsync
 	"github.com/go-kit/kit/log"
 
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
:+1:

hiddeco commented on pkg/chartsync/git.go:
```diff
@@ -281,12 +291,22 @@ func (c *GitChartSync) sync(hr *v1.HelmRelease, mirrorName string, repo *git.Rep
-func (c *GitChartSync) maybeMirror(mirrorName string, remote string) bool {
+func (c *GitChartSync) maybeMirror(mirrorName string, source *v1.GitChartSource, namespace string) bool {
+	gitURL := source.GitURL
+	if source.SecretRef != nil {
```
As you are doing the same `nil` check in `addAuthForHTTPS`, I would reduce this to

```go
gitURL := source.GitURL
var err error
if gitURL, err = c.addAuthForHTTPS(source.GitURL, source.SecretRef, namespace); err != nil {
  c.logger.Log("error", GitAuthError{err}.Error())
  return false
}
```

richardcase replied:
A good suggestion, changed.

hiddeco reviewed (commented):
Besides the in-code comments, I think the `mirrorName` function needs to be made smarter as the credentials from other `HelmRelease` resources (and namespaces) will now leak by configuring a `source.GitURL` (without credentials) that is already in use by another `HelmRelease` (with credentials).

You may get away with just adding the namespace, if I am not mistaken this is the same approach Flux uses for image pull secrets.

richardcase committed:
chore: review changes

Incorporated changes from review. Changed the mirror name to be and
composite key comprising namespace, auth secret name and git url.

richardcase commented:
> You may get away with just adding the namespace, if I am not mistaken this is the same approach Flux uses for image pull secrets.

@hiddeco - i checked and flux does use the namespace. I changed the `mirrorName` to use a composite of namespace, auth secret name, git url. Do you think this is ok?

hiddeco commented on pkg/chartsync/git.go:
```diff
@@ -314,7 +332,60 @@ func (c *GitChartSync) helmReleasesForMirror(mirror string) ([]*v1.HelmRelease,
 func mirrorName(hr *v1.HelmRelease) string {
 	if hr != nil && hr.Spec.GitChartSource != nil {
-		return hr.Spec.GitURL
+		secretName := "noauth"
```
I think this approach is ideologically correct when secrets are being used, but this also means we will do a lot more (duplicate) polling for e.g. releases without HTTPS authentication. My propose would be to change it to something like this:

```go
if hr != nil && hr.Spec.GitChartSource != nil {
	if hr.Spec.GitChartSource.SecretRef == nil {
		return hr.Spec.GitURL
	}
	return fmt.Sprintf("%s/%s/%s", hr.GetNamespace(), hr.Spec.GitChartSource.SecretRef.Name, hr.Spec.GitURL)
}
return ""
```

In addition, the `TODO(michael)` comment can also be removed.

richardcase committed:
feat: changed the mirror name logic

Based on review feedback the mirrot name logic has been changed to
reduce duplicate mirrors if there are no creds

hiddeco commented:
@richardcase thanks a lot :1st_place_medal: 

@stefanprodan can you take a look and see if this matches your expectations for the issue you mentioned? Feature wise it looks fine and mergeable to me (after I get rid of the persistent alias caused by the linter).

stefanprodan commented:
@hiddeco it looks Ok to me, it needs a rebase with master are you going to do that when fixing the aliases?

hiddeco commented:
@stefanprodan yep, will do the rebase and some minor fixes that triggered my OCD :+1:

hiddeco committed:
Small rephrase of method docblocks

hiddeco reviewed (approved):
Rebased and tidied up, awesome contribution @richardcase :raised_hands: :cake:

stefanprodan commented on pkg/chartsync/git.go:
```diff
@@ -284,12 +292,20 @@ func (c *GitChartSync) sync(hr *v1.HelmRelease, mirrorName string, repo *git.Rep
 		git.PollInterval(c.config.GitPollInterval), git.ReadOnly)
 	if !ok {
-		c.logger.Log("info", "started mirroring new remote", "remote", remote, "mirror", mirrorName)
+		c.logger.Log("info", "started mirroring new remote", "remote", source.GitURL, "mirror", mirrorName)
```
Do we log the credentials here?

hiddeco replied:
No, `source.GitURL` does not include the credentials but is the 1:1 copy from the `HelmRelease` definition.

stefanprodan replied:
Great!

stefanprodan commented on pkg/chartsync/git.go:
```diff
@@ -78,11 +85,12 @@ func (c sourceRef) forHelmRelease(hr *v1.HelmRelease) bool {
 	return &GitChartSync{
 		logger:             logger,
 		config:             cfg,
+		coreV1Client: 		coreV1Client,
```
Please run `go fmt`.

stefanprodan commented:
In the future we should consider adding support for cross-namespace secrets like we have done in https://github.com/fluxcd/helm-operator/pull/219

stefanprodan reviewed (approved):
LGTM
